lib/ddtrace/buffer.rb

```diff
@@ -8,38 +8,28 @@ class TraceBuffer
     def initialize(max_size)
       @max_size = max_size
 
-      @mutex = Mutex.new()
+      @mutex = Mutex.new
       @traces = []
       @closed = false
     end
 
     # Add a new ``trace`` in the local queue. This method doesn't block the execution
     # even if the buffer is full. In that case, a random trace is discarded.
     def push(trace)
-      @mutex.synchronize do
-        return if @closed
-        len = @traces.length
-        if len < @max_size || @max_size <= 0
-          @traces << trace
-        else
-          # we should replace a random trace with the new one
-          @traces[rand(len)] = trace
-        end
-      end
+      return if @closed
+      len = @traces.length
+      @traces.delete_at(rand(len)) if len >= @max_size && @max_size > 0
```

Comment:
Reviewing this together, we noted a possible edge case where this allows the buffer to go over `max_size` by a very small amount (say 1 or so) when two threads reach this block at the same time without the mutex. This might be a small sacrifice we want to make for a performance gain.